lib/rubygems/executor.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+require 'open-uri'
+require 'json'
+
+module Gem
+  module Web
+    class Executor
+
+      def open_page(gem, options)
+        begin
+          spec = Gem::Specification.find_by_name(gem)
+        rescue Gem::MissingSpecError => e
+          spec = fetch_remote_spec(gem)
+
+          if spec.nil?
+            puts "Could not find '#{gem}' in rubygems.org."
+            return
+          end
+        end
+
+        if options[:sourcecode]
+          get_info_from_metadata(spec, "source_code_uri")
+        elsif options[:doc]
+          get_info_from_metadata(spec, "documentation_uri")
+        elsif options[:rubygems]
+          open_rubygems(gem)
+        else
+          open_browser(spec.homepage)
+        end
+      end
+
+      def fetch_remote_spec(gem)
+        dep = Gem::Dependency.new(gem)
+        found, _ = Gem::SpecFetcher.fetcher.spec_for_dependency(dep)
+        spec_tuple = found.first
+
+        spec_tuple.first unless spec_tuple.nil? || spec_tuple.empty?
+      end
+
+      def get_info_from_metadata(spec, info)
+        uri = spec.metadata[info]&.strip
+
+        if uri.nil? || uri.empty?
+          puts "Gem '#{spec.name}' does not specify #{info}."
+        else
+          open_browser(uri)
+        end
+      end
+
+      def open_rubygems(gem)
+        open_browser("https://rubygems.org/gems/#{gem}")
```

Comment:
:thinking: 

This is reliable address only for gems coming from `rubygems.org` source. I think to make this command bulletproof it will be safer to remove this auto-generated link and rely only on gemspec info provided by gem author.

Alternatively we should track the origin of gem and use this address only for gems coming from `rubygems.org`.

Reply:
Looking through everything again, I think this is the last remaining problem preventing this from being merged.

It may make sense to rename the `-r`/--rubygems` flag once this is changed, but I'm not sure what it would be changed to.

I'll poke at this a bit and see if I can find anything about getting the origin for the gem.

Reply:
@duckinator Ok! If needed, I can remove the option at all.

Reply:
@MatheusRich Unfortunately, it looks the `.gemspec` files for installed gems don't include any information about their origin. I think it may be best to just remove the `-r`/`--rubygems` flags for the time being.

Someone can always add an equivalent later if a way to get the upstream information is found/added.

Reply:
I think we do not store gem origin. It would be great new feature IMHO, but for now since the option is called `rubygems` I think it makes sense to keep it as is. Anyway I think we can extend option help message to mention this problem.

Reply:
@simi trying to make sure I understood correctly -- are you saying to just keep it `-r`/`--rubygems` but add a note to `gem web --help` saying that it won't work right for gems not hosted on rubygems.org?

Reply:
Yup, my idea is to keep `-r/--rubygems` and explain at `gem web --help` that it just trying to blidly find that gem on `rubygems.org` by name.

Reply:
@MatheusRich I think following @simi's approach (just adding a note to `gem web --help` that `-r`/`--rubygems` always assumes things are hosted on rubygems.org) should be fine. If anyone comes up with a better approach, someone can always make a PR later on.

Reply:
Ok! I'll update it

Reply:
@duckinator What do you think?

Reply:
I think we should mention if gem is not hosted on rubygems it can fail or find another gem with same name hosted on rubygems. Anyway I have no idea how to write that in nice short way (I'm not native english speaker). Anyone any idea?

Reply:
I made a suggestion here: https://github.com/rubygems/rubygems/pull/2835/files#r344308330 :+1: